File: LowestCommonAncestor.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
  static HashMap<TreeNode, TreeNode> map = new HashMap<>();

  static Set<TreeNode> set = new HashSet<>();
  public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
    setUp(root);
    setUpSet(p);


    while (q != null) {
      if (set.contains(q)) {
        return q;
      }
      q = getParent(q);
    }
    return null;
  }

  public void setUp(TreeNode root) {
    if (root == null) {
      return;
    }

    map.put(root.left,  root);
    map.put(root.right, root);

    setUp(root.left);
    setUp(root.right);
  }

  public void setUpSet(TreeNode root) {
    while (root != null) {
      set.add(root);
      root = getParent(root);
    }
  }

  public TreeNode getParent(TreeNode root) {
    return map.get(root);
  }
}

File: LowestCommonAncestor.py
# Definition for a binary tree node.
# class TreeNode(object):
#
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution(object):

    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """

        if (root.val >= p.val and root.val <= q.val) or (root.val <= p.val and root.val >= q.val):
            return root
        if root.val < p.val:
            return self.lowestCommonAncestor(root.right, p, q)
        else:
            return self.lowestCommonAncestor(root.left, p, q)

from TestObjects import *

b = BinarySearchTree()
p = TreeNode(0)
q = TreeNode(5)
s = Solution()
print s.lowestCommonAncestor(b.root, p, q).val
